### PR TITLE
Works in Russian

### PR DESCRIPTION
**Pull request by erikpukinskis**
Rewrote the algorithm, much more verbose now. About 4x slower according to my tests. But should work in any language supported by JavaScript.

¯\\_(ツ)\_/¯

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -7,12 +7,55 @@
...
   }
-  str = str.replace(/([a-z])([A-Z])/g, '$1-$2');
-  str = str.replace(/[ \t\W]/g, '-');
-  str = str.replace(/^-+|-+$/g, '');
-  return str.toLowerCase();
+
+  var sentence = []
+  var needsDash = false
+
+  for(var i=0; i<string.length; i++) {
```
Comment (jonschlinkert) [inline index.js]: Hi @erikpukinskis, thanks for the pr! would you mind fixing code formatting to be consistent with the rest of the code? (whitespace and semicolons mostly. if you're unsure you can run eslint. thanks in advance)

File: index.js [hunk 2]
```diff
@@ -7,12 +7,55 @@
...
+    if (needsDash && isAlphaNumeric) {
+      sentence.push("-");
+      needsDash = false;
+    } else if (isUpper && sentence.length) {
+      sentence.push("-");
+    }
+
+    if (isAlphaNumeric) {
+      if (lower == "*") {
+        throw new Error("daf")
```
Comment (jonschlinkert) [inline index.js]: I'm curious, what does `daf` mean? also, what's the reasoning behind throwing here? `*` is a valid character that should be converted to `-` I believe.

**Comment by erikpukinskis** [comment]
The additional time seems to be about 1 part toUpperCase(), 1 part toLowerCase(), and 1 part parseFloat(). So, the 4x slowdown is pretty much fundamental to this specific approach to solving the problem.

I'm sure there's a faster way, but this way seems very "correct" to me, in that it just uses JavaScript built-ins to decide if a character is upper, lower, or numeric.

If you want to test the speed, you can check out the old version:

    git show 625e9:index.js > old-index.js

And then run this test:

```javascript
var newDashify = require("./index.js")
var oldDashify = require("./old-index.js")

function letters() {
  return Math.random().toString(36)
}

function string() {
  return letters()+" "+letters()+" "+letters()+" "+letters()
}

var strings = []
for(var i=0; i<10000; i++) {
  strings[i] = string()
}

var start = new Date()
strings.map(oldDashify)
var oldTime = new Date() - start

start = new Date()
strings.map(newDashify)
var newTime = new Date() - start

console.log("old version took", oldTime)
console.log("new version took", newTime)
```

File: index.js (hunk 2 above)
Comment (erikpukinskis) [inline index.js]: Ah, debugging code I left in!

File: index.js (hunk 1 above)
Comment (erikpukinskis) [inline index.js]: Got 'em! I don't use semi-colons unless I'm starting a line with an array/object literal, so I tried to add them in, but missed a few. :)

**Comment by jonschlinkert** [comment]
@erikpukinskis thanks for the fast response! fwiw I'm going to be packing to leave on a trip in a few min, if possible I'll review before I leave or while I'm traveling, but it's possible it might be a few days before I can get back to this. Sorry! bad timing on my part!

File: index.js (hunk 1 above)
Comment (erikpukinskis) [inline index.js]: Let me know if there are whitespace changes you want? Not sure how other people like it.

**Comment by erikpukinskis** [comment]
@jonschlinkert That's all good! Enjoy your trip. I've got my own version of dasherize inline'd in my project for temporary, I just didn't want to create yet another NPM module since yours has the same semantics as what I wanted.

**Comment by jonschlinkert** [comment]
(fwiw i'm also trying to publish a lib I've been working on before I leave. In case you see that pushed up, don't think I forgot about this, just trying to prioritize before I leave) thanks again!

**Comment by erikpukinskis** [comment]
Just a reminder to peek at this!

I only thought of it because I went looking for my dasherifizicize code for a new project and I'm too lazy to change my npm files to git and then back again.

If you take too long I'm going to register dasherifizicize on npm and then the world will be an uglier place :)

**Comment by jonschlinkert** [comment]
> If you take too long I'm going to register dasherifizicize

haha none of us want to see that happen. sorry for the lack of response, I'll pull this down and review this weekend. thanks for the ping!

**Comment by erikpukinskis** [comment]
I would love you so hard if you clicked that merge button, and published to npm.

For now, this is my life:

    cp ../meals/dashify.js .

**Comment by erikpukinskis** [comment]
Truth be told, I'm starting to wonder if it's not better to copy/paste this kind of stuff anyway! After the left-pad fiasco, and [these kinds of security issues](https://twitter.com/iamakulov/status/892485192883073024)... I'm a little wary of npm dependencies. But that's part of our reality I guess.

**Comment by jonschlinkert** [comment]
@erikpukinskis sorry for the late reply. It seems like we should have been able to just update the regex to fix this issue. is there a reason why regex wouldn't work?

**Comment by erikpukinskis** [comment]
I’m not aware of another way to tell alphabetical characters from non-alphabet characters besides my upper/lowercase hack. And even that doesn’t work with all alphabets.

I guess you could maybe test for \w which gives you letters, numbers, and underscore, and then detect out the numbers and underscore. Might be worth a shot.

**Pull request closed without merging**